share/functions/__fish_print_hostnames.fish

```diff
@@ -66,25 +65,24 @@ function __fish_print_hostnames -d "Print a list of known hostnames"
                 end
             end
 
-            # Skip unusable paths.
-            test -d "$relative_path" -a -x "$relative_path"
-            or return
-
-            builtin cd $relative_path
             set -l new_paths
             for path in $paths
                 set -l expanded_path
-                eval "set expanded_path (printf \"%s\n\" $path)"
+                # Scope "relative" paths in accordance to ssh path resolution
+                if string match -qrv '^[~/]' $path
+                    set path $relative_path/$path
+                end
+                # Expand paths (eg ~/.ssh/../test/* to /home/<user>/test/file1 /home/<user>/test/file2)
+                # Use eval to prevent "No matches for wildcard" messages
+                eval set expanded_path (echo $path)
```

Comment:
That comment is incorrect. Without `eval`, no wildcard expansion will take place here!

And with `set`, the error should not be printed.

So there's something else going on here.

Reply:
It is true in 2.7.1. If it is not true for this version (master), then I can try and do two separate PRs. I can build the master branch but I cannot get it running so I might need some help for a master PR.

Reply:
If you do e.g. `set -l path '*'` and then `set -l expanded_path (echo $path)`, then `$expanded_path` will still be a literal `*`. If you use `eval`, that `*` will be interpreted as a wildcard. Since you use that wildcard with the `set` command, no error for a missing match will be printed (that error is disabled for `for`, `set` and `count`).

That's true for 2.7.1 as well (I did test it!), so I'm not quite sure why you experience something different.

Try this:

```fish
$ ls *thisshouldnotmatch
# fish errors out because the wildcard did not match
$ ls '*thisshouldnotmatch'
# ls complains about file or directory not found
$ eval ls '*thisshouldnotmatch'
# fish errors out because there's no match, `ls` isn't executed
$ set -l var '*thisshouldnotmatch'
# no output
$ ls $var
# same as the ls above
$ eval ls $var
# another fish error
$ eval set othervar $var
# no error, $othervar is empty
```

Most likely the eval is correct here (though a safer alternative would be much nicer!), and the comment is just wrong.

Reply:
My apologies. You are right. I misread your feedback. I was able to reduce the line to `eval set expanded_path $path`. The line I am replacing `eval "set expanded_path (printf \"%s\n\" $path)"` and the 2.7.1 version `echo "set expanded_path (printf \"%s\n\" $path)" | source` does give me the wildcard error though. I could only test those lines on 2.7.1.

> though a safer alternative would be much nicer!
Agreed. I really wish there was a path resolver function which includes globbing. The current approach is all I know.

> and the comment is just wrong.
I will fix the comment in the next few days.